devito/finite_differences/coefficients.py

```diff
@@ -0,0 +1,66 @@
+import sympy
+import numpy as np
+
+from cached_property import cached_property
+
+from devito.tools import filter_ordered, flatten
+
+__all__ = ['Coefficients']
+
+
+class Coefficients(object):
+    """
+    Devito class for users to define custom finite difference weights.
+    """
+ 
+    def __init__(self, *args, **kwargs):
+
+        self.check_args(args, kwargs)
+
+        self.data = args
+        self.rules = self.rules()
+        
+    def check_args(self, args, kwargs):
+        for arg in args:
+            if isinstance(arg, tuple):
+                assert isinstance(arg[0], int)
+                assert(arg[1].is_Function)
+                assert(arg[2].is_Dimension)
+                assert isinstance(arg[3], np.ndarray)
+            else:
+                raise NotImplementedError
```

Comment:
wouldn't it be nicer and "more readable" to have `__init__(order, function, dimensions, coeffs, **kwargs)` rather than obscure `args` with a bunch of tests with it?

Reply:
Depending on how symbolic coefficients are going to be used, the input could be in various formats i.e. grid wide (which will be introduced in this PR) vs. modified at specific nodes vs. grid wide function (future PR's). 

So I'll probably need to iterate this a few times but however it develops we need to keeps it flexible in order that future functionality can be incorporated without too much messing around.

Reply:
I understand that, but I still think the `__init__` signature should be explicit enough for people to know what it is like we did for most of the Devito codebase (Function,...).